### PR TITLE
Fix perftuning to meet DLIS integration requirements

### PR DESCRIPTION
**Pull request by guotuofeng**
## Describe your changes

* Change the execution provider priority to use the parameters instead of that one in inference_settings. 
    Bug description: In DLIS scenarios, the inferenece_settings is something like:
{'execution_provider': [('MIGraphXExecutionProvider', {})]} and the provider_lists specified in perf_tuning is: ['ROCMExecutionProvider', 'MIGraphXExecutionProvider'].

    Originally, the inference_settings priority is higher than the execution_providers specified in arguments. As the result, when calling OnnxEvaluator.evaluate with the provider_list in baseline evaluation, the execution provider will be used as ["MIGraphXExecutionProvider"]. Later, the OnnxEvaluator.disable_ort_fallback(session, execution_providers) will check whether the underlying session's providers(MIGraphXExecutionProvider, CPUExecutionProvider) contains the execution_providers(ROCMExecutionProvider, MIGraphXExecutionProvider). In this case, the ROCMExecutionProvider is excluded when creating inference session. and the OliveEvaluationError exception will be raised.

    To fix this issue, we need move the check_ort_fallback into prepare_session in onnx model.
    perf tuning will always pass None as the execution provider to evaluate to make the logic clear

* Change perf tuning to always has execution provider info regardless of whether the best result is baseline or from perf tuning.
* Change perf tuning to accept both provider list and provider list with provider options.
* Add force_evaluate_other_eps to force run all other EPs if the current EP is not same with the additional EPs
* Change OnnxModel.prepare_session to accept both execution provider list and (execution provider, provider options) list
In this way, the EP/provider option handling logic is normalized.
* Add MIGraphXExecutionProvider to the accelerator map
* Fix the critical bug when creating inference session that doesn't respect provider options in ort_inference.py
* Add io_bind inference for accuracy since illegal memory access in cuda will be thrown if enable_cuda_graph. 
  If io_bind is not enabled and enable_cuda_graph is True, the following errors will be raised:

    onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : CUDA failure 700: an illegal memory access was encountered ; GPU=0 ; hostname=c93f1847c000000 ; file=/onnxruntime_src/onnxruntime/core/providers/cuda/cuda_graph.cc ; line=49 ; expr=cudaGraphLaunch(graph_exec_, stream_); 
    [E:onnxruntime:Default, cuda_call.cc:116 CudaCall] CUDA failure 700: an illegal memory access was encountered ; GPU=0 ; hostname=c93f1847c000000 ; file=/onnxruntime_src/onnxruntime/core/providers/cuda/cuda_execution_provider.cc ; line=286 ; expr=cudaStreamDestroy(stream_); 
    [E:onnxruntime:Default, cuda_call.cc:116 CudaCall] CUDNN failure 4: CUDNN_STATUS_INTERNAL_ERROR ; GPU=0 ; hostname=c93f1847c000000 ; file=/onnxruntime_src/onnxruntime/core/providers/cuda/cuda_execution_provider.cc ; line=181 ; expr=cudnnDestroy(cudnn_handle_); 

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/model/handler/onnx.py [hunk 1]
```diff
@@ -74,28 +74,55 @@ def prepare_session(
...
+        else:
+            execution_providers = inference_settings.get("execution_provider")
+            provider_options = inference_settings.get("provider_options")
+
         if not execution_providers:
             execution_providers = self.get_default_execution_providers(device)
-        elif isinstance(execution_providers, str):
+            provider_options = None
+        elif isinstance(execution_providers, (str, tuple)):
             execution_providers = [execution_providers]
```
Comment (trajepl) [inline olive/model/handler/onnx.py:108]: Maybe more clear? 
```python

        if execution_providers:
            provider_options = None
        else:
            execution_providers = inference_settings.get("execution_provider")
            provider_options = inference_settings.get("provider_options")
            if not execution_providers:
                execution_providers = self.get_default_execution_providers(device)
                provider_options = None

        if isinstance(execution_providers, (str, tuple)):
            execution_providers = [execution_providers]


```
(BTW, can we assume that if the execution_providers is None from inference_settings, then `provider_options` must be None?)

File: olive/model/utils/onnx_utils.py [hunk 2]
```diff
@@ -67,3 +71,111 @@ def get_onnx_file_path(model_path: str, onnx_file_name: Optional[str] = None) ->
...
     else:
         raise ValueError(f"No .onnx file found in the model folder {model_path}.")
+
+
+def check_and_normalize_provider_args(
+    providers: Sequence[Union[str, Tuple[str, Dict[Any, Any]]]],
+    provider_options: Sequence[Dict[Any, Any]],
+    available_provider_names: Sequence[str],
+):
+    """Validate the 'providers' and 'provider_options' arguments and returns a normalized version.
```
Comment (trajepl) [inline olive/model/utils/onnx_utils.py:81]: Does `copy but not call from ort` because this function only exists after specific ort version?

File: test/unit_test/passes/onnx/test_perf_tuning.py [hunk 3]
```diff
@@ -61,6 +65,113 @@ def test_ort_perf_tuning_with_customized_configs(mock_run, config):
...
+    if return_baseline:
+        assert len(acutal_eps) == 2
+        assert "enable_cuda_graph" not in acutal_eps[0][1]
+
+        assert acutal_eps[1] == "CPUExecutionProvider"
+        assert f"Best result({PERFTUNING_BASELINE}):" in caplog.text
+    else:
+        assert len(acutal_eps) == 1
+        if acutal_eps[0][0] == "CUDAExecutionProvider":
+            assert "enable_cuda_graph" in acutal_eps[0][1]
```
[the remark this answers is not in the document]
Comment (guotuofeng) [reply test/unit_test/passes/onnx/test_perf_tuning.py:121]: I don't run evaluator. I just mock the evaluate.

File: test/unit_test/model/test_onnx_model.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        "session_options": {
+            "execution_mode": 0,
+            "graph_optimization_level": 99,
+            "extra_session_config": None,
+            "inter_op_num_threads": None,
+            "intra_op_num_threads": 1,
+        },
+    }
+
+    execution_providers = ["ROCMExecutionProvider", "MIGraphXExecutionProvider", "CUDAExecutionProvider"]
```
[the remark this answers is not in the document]
Comment (guotuofeng) [reply test/unit_test/model/test_onnx_model.py:30]: It is just for UT

File: olive/model/utils/onnx_utils.py (hunk 2 above)
Comment (guotuofeng) [reply olive/model/utils/onnx_utils.py:81]: This function is not exposed. I copy here just avoid the signature changes breaking Olive in future.

File: olive/model/handler/onnx.py (hunk 1 above)
Comment (guotuofeng) [reply olive/model/handler/onnx.py:108]: will do

**Review by trajepl — APPROVED** [review]

File: olive/model/handler/onnx.py [hunk 5]
```diff
@@ -74,29 +74,58 @@ def prepare_session(
...
-            ]
+        # split the execution_providers and provider_options
+        execution_providers, provider_options = check_and_normalize_provider_args(
+            execution_providers, provider_options, ort.get_available_providers()
+        )
 
-        return get_ort_inference_session(self.model_path, inference_settings, self.use_ort_extensions)
+        if (device == Device.GPU) and (rank is not None):
+            for i, ep in enumerate(execution_providers):
+                if ep == "CUDAExecutionProvider" and not provider_options[i]:
```
Comment (jambayk) [inline olive/model/handler/onnx.py:122]: Wouldn't this condition return false if provider options is `{}`? Don't we want to provide the `device_id` even when it is `{}`?

Comment (jambayk) [reply olive/model/handler/onnx.py:122]: Actually, never mind. I misread the condition. This is true only when provider options is `{}`. So, we want to insert the device id if the user hasn't specified the provider options.

Comment (jambayk) [reply olive/model/handler/onnx.py:122]: maybe we could instead check for `provider_options.get(device_id) is not None` since for distributed models, we want to have device id always?

File: olive/model/handler/onnx.py [hunk 6]
```diff
@@ -74,29 +74,58 @@ def prepare_session(
...
+        # Originally, the inference_settings priority is higher than the execution_providers specified in arguments.
+        # as the result, when calling OnnxEvaluator.evaluate with the provider_list in baseline evaluation,
+        # the execution provider will be used as ["MIGraphXExecutionProvider"].
+        # later, the check_ort_fallback(session, execution_providers) will check the
+        # whether the underlying session's providers(MIGraphXExecutionProvider, CPUExecutionProvider) contains the
+        # execution_providers(ROCMExecutionProvider, MIGraphXExecutionProvider). In this case, the
+        # ROCMExecutionProvider is excluded when creating inference session. and the OliveEvaluationError exception
+        # will be raised.
+        # To fix this issue, we need to make sure the execution_providers specified in arguments has higher priority
+        if execution_providers:
```
Comment (jambayk) [inline olive/model/handler/onnx.py:98]: Olive evaluator always provides execution provider https://github.com/microsoft/Olive/blob/3a774a75592cd4d12c5836dea19d6d462ccbaeac/olive/systems/local.py#L49 

So, wouldn't this ignore the perf tuned inference settings when evaluating the output model of the perf-tuning pass?

Comment (jambayk) [reply olive/model/handler/onnx.py:98]: Or it doesn't matter since only the `provider_options` is ignored in this case and we don't tune it. But what about settings like `trt_fp16_enable` and `enable_cuda_graph`? https://github.com/microsoft/Olive/blob/3a774a75592cd4d12c5836dea19d6d462ccbaeac/olive/passes/onnx/perf_tuning.py#L135 

Perhaps, it would be more complete if we resolve the `ep, provider_options` for both the user provided `ep` and model's inference settings. As use the inference setting's provider options for eps from user that don't have their own provider options?

File: olive/model/handler/onnx.py (hunk 5 above)
Comment (guotuofeng) [reply olive/model/handler/onnx.py:122]: This is the original behavior
https://github.com/microsoft/Olive/blob/main/olive/model/handler/onnx.py#L94-L97

File: olive/model/handler/onnx.py (hunk 6 above)
Comment (guotuofeng) [reply olive/model/handler/onnx.py]: fixed.

**Review by jambayk — APPROVED** [review]

**Review by jambayk — DISMISSED** [review]

File: olive/evaluator/olive_evaluator.py [hunk 7]
```diff
@@ -468,19 +430,53 @@ def _inference(
...
         preds = []
         targets = []
         logits = []
         logits_dict = collections.defaultdict(list)
         output_names = io_config["output_names"]
+        io_bind = self.io_bind_enabled(metric, model.inference_settings)
+        device = "cuda" if device == "gpu" else "cpu"
+        if io_bind and device == "cuda":
+            io_binding = session.io_binding()
+            input_data, _ = next(iter(dataloader))
```
Comment (trajepl) [inline olive/evaluator/olive_evaluator.py:444]: We may need to rewind the dataloader after `next(iter(dataloader))`.

Otherwise, the next `for input_data, labels in dataloader` will start from second one of the full data.

Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:444]: I don't think that's needed? We expect the dataloader to be similar to torch dataloader. There is no concept of rewind like there is for the onnxruntime quantization calibration dataloader. 

For this dataloader, the expectation is that every new `iter(dataloader)` call on the dataloader starts the iteration from the start. 
The for loop is also an instance of a new `iter` over the dataloader.

Comment (trajepl) [reply olive/evaluator/olive_evaluator.py:444]: I am wrong.  No need to remind.

**Review by trajepl — APPROVED** [review]
